content.go

```diff
@@ -82,60 +82,43 @@ func TagN(ctx context.Context, target Target, srcReference string, dstReferences
 		opts.MaxMetadataBytes = defaultTagNMaxMetadataBytes
 	}
 
-	refFetcher, okFetch := target.(registry.ReferenceFetcher)
-	refPusher, okPush := target.(registry.ReferencePusher)
-	if okFetch && okPush {
+	_, isRefFetcher := target.(registry.ReferenceFetcher)
+	_, isRefPusher := target.(registry.ReferencePusher)
+	if isRefFetcher && isRefPusher {
 		if repo, ok := target.(interfaces.ReferenceParser); ok {
 			// add scope hints to minimize the number of auth requests
 			ref, err := repo.ParseReference(srcReference)
 			if err != nil {
-				return err
+				return ocispec.Descriptor{}, err
 			}
 			ctx = registryutil.WithScopeHint(ctx, ref, auth.ActionPull, auth.ActionPush)
 		}
 
-		desc, contentBytes, err := func() (ocispec.Descriptor, []byte, error) {
-			desc, rc, err := refFetcher.FetchReference(ctx, srcReference)
-			if err != nil {
-				return ocispec.Descriptor{}, nil, err
-			}
-			defer rc.Close()
-
-			if desc.Size > opts.MaxMetadataBytes {
-				return ocispec.Descriptor{}, nil, fmt.Errorf(
+		desc, contentBytes, err := FetchBytes(ctx, target, srcReference, FetchBytesOptions{
+			MaxBytes: opts.MaxMetadataBytes,
+		})
+		if err != nil {
+			if errors.Is(err, errdef.ErrSizeExceedsLimit) {
+				err = fmt.Errorf(
 					"content size %v exceeds MaxMetadataBytes %v: %w",
 					desc.Size,
 					opts.MaxMetadataBytes,
 					errdef.ErrSizeExceedsLimit)
 			}
-			contentBytes, err := content.ReadAll(rc, desc)
-			if err != nil {
-				return ocispec.Descriptor{}, nil, err
-			}
-			return desc, contentBytes, nil
-		}()
-		if err != nil {
-			return err
+			return ocispec.Descriptor{}, err
 		}
 
-		eg, egCtx := syncutil.LimitGroup(ctx, opts.Concurrency)
-		for _, dstRef := range dstReferences {
-			eg.Go(func(dst string) func() error {
-				return func() error {
-					r := bytes.NewReader(contentBytes)
-					if err := refPusher.PushReference(egCtx, desc, r, dst); err != nil && !errors.Is(err, errdef.ErrAlreadyExists) {
-						return fmt.Errorf("failed to tag %s as %s: %w", srcReference, dst, err)
-					}
-					return nil
-				}
-			}(dstRef))
+		if err := tagBytesN(ctx, target, desc, contentBytes, dstReferences, TagBytesNOptions{
```

Comment:
Can we remove the reference fetcher and reference pusher check and always use `content.FetchBytes` and `content.tagBytesN`?

Reply:
Please resolve it. Looks like the proposal doesn't perform well when the Target is local, e.g. memory target.